contracts/upgradeable_contracts/BridgeValidators.sol

```diff
@@ -9,89 +9,167 @@ import "../upgradeability/EternalStorage.sol";
         }
+
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         uintStorage[keccak256("deployedAtBlock")] = block.number;
         setInitialize(true);
         emit RequiredSignaturesChanged(_requiredSignatures);
+
         return isInitialized();
     }
 
-    function addValidator(address _validator) external onlyOwner {
-        require(_validator != address(0));
+    function addValidator(address _validator, address _reward) external onlyOwner {
+        require(_validator != address(0) && _validator != F_ADDR);
+        require(_reward != address(0));
         require(!isValidator(_validator));
+
+        address firstValidator = getNextValidator(F_ADDR);
+        setValidator(_validator, firstValidator);
+        setValidatorRewardAddress(_validator, _reward);
+        setValidator(F_ADDR, _validator);
+
         setValidatorCount(validatorCount().add(1));
-        setValidator(_validator, true);
         emit ValidatorAdded(_validator);
     }
 
     function removeValidator(address _validator) external onlyOwner {
         require(validatorCount() > requiredSignatures());
         require(isValidator(_validator));
-        setValidator(_validator, false);
+        address validatorsNext = getNextValidator(_validator);
+        address index = F_ADDR;
+        address next = getNextValidator(index);
+
+        // find the element in the list pointing to _validator
+        while (next != _validator) {
+            index = next;
+            next = getNextValidator(index);
+        }
+
+        setValidator(index, validatorsNext);
+        deleteItemFromAddressStorage("validatorsList", _validator);
+        deleteItemFromAddressStorage("validatorsRewards", _validator);
         setValidatorCount(validatorCount().sub(1));
+
         emit ValidatorRemoved(_validator);
     }
 
-    function setRequiredSignatures(uint256 _requiredSignatures) external onlyOwner {
+    function setRequiredSignatures(uint256 _requiredSignatures)
+        external
+        onlyOwner
+    {
         require(validatorCount() >= _requiredSignatures);
         require(_requiredSignatures != 0);
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         emit RequiredSignaturesChanged(_requiredSignatures);
     }
 
-    function getBridgeValidatorsInterfacesVersion() public pure returns(uint64 major, uint64 minor, uint64 patch) {
+    function getBridgeValidatorsInterfacesVersion()
+        public
+        pure
+        returns (uint64 major, uint64 minor, uint64 patch)
+    {
         return (2, 0, 0);
     }
 
-    function requiredSignatures() public view returns(uint256) {
+    function requiredSignatures() public view returns (uint256) {
         return uintStorage[keccak256(abi.encodePacked("requiredSignatures"))];
     }
 
-    function validatorCount() public view returns(uint256) {
+    function validatorCount() public view returns (uint256) {
```

Comment:
In which scenarios is it being used? Could we remove `validatorCount` at all and re-work the corresponding pieces of code to use `getNextValidator` instead similar to the `while`-loop in the `validatorList` method.

Reply:
I don't think `validatorCount` is being used externally (only on unit tests). Here are the uses inside the contract:

https://github.com/poanetwork/poa-bridge-contracts/blob/5c553e06eb323e77c3628549c38f04bf224e07a0/contracts/upgradeable_contracts/BaseBridgeValidators.sol#L15-L19

https://github.com/poanetwork/poa-bridge-contracts/blob/5c553e06eb323e77c3628549c38f04bf224e07a0/contracts/upgradeable_contracts/BridgeValidators.sol#L65-L66

https://github.com/poanetwork/poa-bridge-contracts/blob/5c553e06eb323e77c3628549c38f04bf224e07a0/contracts/upgradeable_contracts/BaseBridgeValidators.sol#L57-L58

For `setRequiredSignatures` and `removeValidator`, `validatorCount` could be refactored to  use `getNextValidator` with the `while` loop and so avoid storing the value.

The only case that worries me is on `validatorList` method. To create the Array to return the list of validators we need to know the size of the Array. Performing the refactor, will cause that by invoking `validatorList` we will be iterating two times on the linked-list, one to get the length and the other to store the validator addresses on the Array to be returned. Also, `validatorList()` method will be used every time a transaction is processed by `fee contract` to distribute the fees.
Perhaps there is a way to avoid iterating the two times on the linked-list, any thoughts on this?

Reply:
Usage a variable in the `validatorList` makes sense since this method will be used intensively. So, let leave `validatorCount` as it is now. Thanks for clarification!!!